src/apps/basicnat/basicnat.lua

```diff
@@ -125,23 +128,71 @@ end
 -- TCP, UDP and ICMP header checksums are translated. For inbound
 -- packets, the destination IP address and the checksums as listed above
 -- are translated.
-local function basic_rewrite(pkt, external_ip, internal_ip, mask)
-   -- Only attempt to alter ipv4 packets. Assume an Ethernet encapsulation.
-   if pkt.data[12] ~= 8 or pkt.data[13] ~= 0 then return pkt end
-   local src_ip, dst_ip = get_src_ip(pkt), get_dst_ip(pkt)
-   if src_ip == external_ip then
-      set_src_ip(pkt, internal_ip)
+
+-- FIXME: Would be nice to have &ip src as an addressable, so we could
+-- pass the address at which to munge as an argument to the handlers
+-- without assuming a certain encapsulation.
+
+-- todo: add comment syntax, "else"
+local dispatch_template = [[
+(incoming_tcp, incoming_udp, incoming_other,
+ outgoing_tcp, outgoing_udp, outgoing_other) => {
+  ip src $external_ip => {
+    tcp => incoming_tcp(&ip[0], &ip_payload[0])
```

Comment:
This ip_payload is a language extension that I haven't formalized yet.  It's like tcp[0] but without requiring that it be the first packet in the stream.